src/lib/shim/build.rs

```diff
@@ -0,0 +1,104 @@
+use std::collections::HashMap;
+use std::env;
+use std::path::{Path, PathBuf};
+
+use shadow_build_common::ShadowBuildCommon;
+
+fn run_bindgen(build_common: &ShadowBuildCommon) {
+    let bindings = build_common
+        .bindgen_builder()
+        .header("shim.h")
+        .allowlist_function("shim_.*")
+        .header("shim_api_c.h")
+        .allowlist_function("shimc_.*")
+        // get libc types from libc crate
+        .blocklist_type("addrinfo")
+        .raw_line("use libc::addrinfo;")
+        .blocklist_type("ifaddrs")
+        .raw_line("use libc::ifaddrs;")
+        .blocklist_type("socklen_t")
+        .blocklist_type("__socklen_t")
+        // Finish the builder and generate the bindings.
+        .generate()
+        // Unwrap the Result and panic on failure.
+        .expect("Unable to generate bindings");
+    // Write the bindings to the $OUT_DIR/bindings.rs file.
+    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
+    bindings
+        .write_to_file(out_path.join("bindings.rs"))
+        .expect("Couldn't write bindings!");
+}
+
+fn run_cbindgen(build_common: &ShadowBuildCommon) {
+    let base_config = build_common.cbindgen_base_config();
+    let crate_dir = std::env::var("CARGO_MANIFEST_DIR").unwrap();
+
+    let config = cbindgen::Config {
+        sys_includes: vec![
+            "ifaddrs.h".into(),
+            "netdb.h".into(),
+            "stdarg.h".into(),
+            "sys/socket.h".into(),
+            "sys/types.h".into(),
+            "netdb.h".into(),
+        ],
+        after_includes: {
+            let mut v = base_config.after_includes.clone().unwrap_or_default();
+            // We have to manually create the vararg declaration.
+            v.push_str("long shim_api_syscall(long n, ...);\n");
```

Comment:
Is there any guarantee that a vararg function has the same calling convention as a non-vararg function? And probably worth adding a comment in the bindings that passing more than 7 args will invoke UB (not that I expect anyone to here, but I think it's worth documenting when the function prototype doesn't match the function definition).

Reply:
> Is there any guarantee that a vararg function has the same calling convention as a non-vararg function?

I haven't been able to find super authoritative documentation to that effect, but we've used it elsewhere. Here's a somewhat related reference: https://www.gnu.org/software/libc/manual/html_node/Why-Variadic.html :

> Some functions such as open take a fixed set of arguments, but occasionally ignore the last few. Strict adherence to ISO C requires these functions to be defined as variadic; in practice, however, the GNU C compiler and most other C compilers let you define such a function to take a fixed set of arguments—the most it can ever use—and then only declare the function as variadic (or not declare its arguments at all!). 

> And probably worth adding a comment in the bindings that passing more than 7 args will invoke UB

Similar to above; this isn't UB. Extra arguments are just ignored.